File: src/harness/approvalLog.ts
import * as fs from "fs";
import * as path from "path";
import * as os from "os";
import { sessionManager } from "../session/manager.js";

export type ApprovalDecision = "approve" | "approve_session" | "deny" | "skip";

export interface ApprovalLogEntry {
  timestamp: string;
  tool: string;
  args: unknown;
  decision: ApprovalDecision;
  sessionId?: string;
}

function getLogPath(): string {
  const dir = path.join(os.homedir(), ".chorus");
  fs.mkdirSync(dir, { recursive: true });
  return path.join(dir, "approval-log.ndjson");
}

export function appendApprovalLog(entry: Omit<ApprovalLogEntry, "timestamp" | "sessionId">): void {
  const full: ApprovalLogEntry = {
    ...entry,
    timestamp: new Date().toISOString(),
    sessionId: sessionManager.getCurrent()?.id,
  };
  try {
    fs.appendFileSync(getLogPath(), JSON.stringify(full) + "\n", "utf-8");
  } catch { /* never crash on audit logging */ }
}

export function readApprovalLog(limit = 50): ApprovalLogEntry[] {
  try {
    const raw = fs.readFileSync(getLogPath(), "utf-8");
    const lines = raw.trim().split("\n").filter(Boolean);
    return lines
      .slice(-limit)
      .map((line) => {
        try { return JSON.parse(line) as ApprovalLogEntry; } catch { return null; }
      })
      .filter((e): e is ApprovalLogEntry => e !== null);
  } catch {
    return [];
  }
}

File: chorus-engine-harness/src/main/java/com/chorus/engine/harness/ApprovalLog.java
package com.chorus.engine.harness;

import org.jspecify.annotations.NonNull;
import org.jspecify.annotations.Nullable;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

/**
 * Append-only audit log for approval decisions.
 * Stores entries as NDJSON (newline-delimited JSON) for durability and easy streaming.
 */
public final class ApprovalLog {

    private final Path logPath;
    private final ObjectMapper mapper;
    private final String sessionId;

    public ApprovalLog(@NonNull Path logPath, @Nullable String sessionId) {
        this.logPath = logPath;
        this.sessionId = sessionId;
        this.mapper = new ObjectMapper().registerModule(new JavaTimeModule());
    }

    /**
     * Append a new approval decision to the log.
     * Never throws — silently drops on I/O failure to avoid blocking execution.
     */
    public void append(
        @NonNull String tool,
        @NonNull Map<String, Object> args,
        ApprovalLogEntry.ApprovalDecision decision
    ) {
        ApprovalLogEntry entry = new ApprovalLogEntry(
            Instant.now(), tool, args, decision, sessionId
        );
        try {
            Files.createDirectories(logPath.getParent());
            String line = mapper.writeValueAsString(entry) + "\n";
            Files.writeString(logPath, line, StandardOpenOption.CREATE, StandardOpenOption.APPEND);
        } catch (IOException e) {
            // Never crash on audit logging
        }
    }

    /**
     * Read the last N entries from the log.
     */
    public @NonNull List<ApprovalLogEntry> readLast(int limit) {
        if (!Files.exists(logPath)) {
            return List.of();
        }
        try {
            List<String> lines = Files.readAllLines(logPath);
            List<ApprovalLogEntry> entries = new ArrayList<>();
            int start = Math.max(0, lines.size() - limit);
            for (int i = start; i < lines.size(); i++) {
                String line = lines.get(i).trim();
                if (line.isEmpty()) continue;
                try {
                    entries.add(mapper.readValue(line, new TypeReference<>() {}));
                } catch (IOException ignored) {
                    // Skip corrupted lines
                }
            }
            return Collections.unmodifiableList(entries);
        } catch (IOException e) {
            return List.of();
        }
    }

    /**
     * Read all entries from the log.
     */
    public @NonNull List<ApprovalLogEntry> readAll() {
        return readLast(Integer.MAX_VALUE);
    }

    /**
     * Count total entries in the log.
     */
    public long count() {
        if (!Files.exists(logPath)) return 0;
        try {
            return Files.lines(logPath).filter(l -> !l.isBlank()).count();
        } catch (IOException e) {
            return 0;
        }
    }
}
